daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SExpr.scala

```diff
@@ -135,9 +135,33 @@ object SExpr {
     }
   }
 
+  // A fully saturated builtin application
+  final case class SEAppB(builtin: SBuiltin, args: Array[SExpr])
+      extends SExpr
+      with SomeArrayEquals {
+    if (args.size != builtin.arity) {
+      throw SErrorCrash(s"SEAppB: arg.size != builtin.arity")
+    }
+    def execute(machine: Machine): Unit = {
+      val arity = builtin.arity
+      val actuals = new util.ArrayList[SValue](arity)
+      machine.pushKont(KBuiltin(builtin, actuals))
+      // Start evaluating the arguments.
+      var i = 1
+      while (i < arity) {
+        val arg = args(arity - i)
+        machine.pushKont(KPushTo(actuals, arg, machine.frame, machine.actuals, machine.env.size))
+        i = i + 1
+      }
+      machine.ctrl = args(0)
+    }
+  }
+
   object SEApp {
     def apply(fun: SExpr, args: Array[SExpr]): SExpr = {
       fun match {
+        case SEBuiltin(builtin) if optimizeAtomicApps && builtin.arity == args.length =>
```

Comment:
Can't we also use this for over-applied built-ins by splitting off the required number of arguments and putting the rest into a separate `SEApp`? That's what would happen at runtime anyway, isn't it?

I was also wondering if actually get the over-applied built-ins but I suppose looking up things in a map-like data structure full of functions and applying arguments to the result might be a candidate, at least in the future.

Reply:
This is indeed what would happen at runtime.
Not a common case I'm sure.
But, it should be easy enough to code, so we can do it...

Reply:
done